sympy/ntheory/qs.py

```diff
@@ -81,43 +83,61 @@ def _generate_factor_base(prime_bound, n):
     """
     from sympy import sieve
     factor_base = []
+    idx_1000, idx_5000 = None, None
     for prime in sieve.primerange(1, prime_bound):
         if pow(n, (prime - 1) // 2, prime) == 1:
+            if prime > 1000 and idx_1000 is None:
+                idx_1000 = len(factor_base) - 1
+            if prime > 5000 and idx_5000 is None:
+                idx_5000 = len(factor_base) - 1
             residue = _sqrt_mod_prime_power(n, prime, 1)[0]
             log_p = round(log(prime)*2**10)
             factor_base.append(FactorBaseElem(prime, residue, log_p))
-    return factor_base
+    return idx_1000, idx_5000, factor_base
 
 
-def _initialize_first_polynomial(N, M, factor_base):
+def _initialize_first_polynomial(N, M, factor_base, idx_1000, idx_5000):
     """This step is the initialization of the 1st sieve polynomial.
     Here `a` is selected as a product of several primes of the factor_base
     such that `a` is about to sqrt(2*N) / M. Other initial values of
     factor_base elem are also intialized which includes a_inv, b_ainv, soln1,
     soln2 which are used when the sieve polynomial is changed. The b_ainv
     is required for fast polynomial change as we don't have to calculate
     `2*b*mod_inverse(a, prime)` every time.
+    We also ensure that the factor_base primes which make `a` are between
```

Comment:
(Here and everywhere)`factor_base` -> variable names should have a single backtick around and mathematical formula/eqs should have double backticks.